### ISSUE TITLE
Emit errors or wranings to user about recursion of display

### ISSUE THREAD

**Issue by gftea** [rust-lang/rust#125292]
### Code

```Rust
This code cause recursion stack overflow

struct Hello;

impl std::fmt::Debug for Hello {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    write!(f, "{}", self)
  }
}

impl std::fmt::Display for Hello {
  fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
    write!(f, "{:8}", self)
  }
}

fn main() {
    let h = Hello;
    println!("{}", h);
}
```

### Current output

```Shell
Runtime error instead of compile time error at present 


thread 'main' has overflowed its stack
fatal runtime error: stack overflow
```
```


### Desired output

```Shell
In a large code base, such small mistake is hard to locate because it does not point out which recursion call cause stack overflow.
```

### Rationale and extra context

It is calling Display in a implementation for Display, this seems to be abvious recursion can be detected by compiler. I guess there may be general effort to detect recursion call, but wonder if such obvious recursion can be prevented by compiler earlier

### Rust Version

```Shell
rustc v1.78.0
```

### Anything else?

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1d2c1c17c729dd7f4be92fd1c2ee14ea

**Comment by workingjubilee** [comment]
Macros, like the one you call there, are pretty good at obscuring these impl details, but yeah we should probably warn.

**Comment by estebank** [comment]
I could have sworn there was a similar report about using `to_string` within an `impl Display for` from maybe 2017/2018, but can't find it 😅 

@workingjubilee in the wild I've seen that simple cases where the chain repeats after just one or two jumps represent the bulk of them, so beyond the obfuscation issues, I'm convincing myself that a very limited call-graph analysis to look for unconditional recursivity is warranted. It's funny that it's always involving `Display` or `Debug` though 😄

**Comment by so-schen** [comment]
And regarding the 2nd suggestion, what is your view

> there is no use case for using `:` as standalone formatter, and it is better the compiler to emit error to use : as standalone formatter because it is likely user want to use `:?` but miss `?` by mistake,

**Comment by tbu-** [comment]
> > there is no use case for using `:` as standalone formatter, and it is better the compiler to emit error to use : as standalone formatter because it is likely user want to use `:?` but miss `?` by mistake,

Unlikely to work, there's probably a crate out there that uses it, so it can't be removed due to backward compatibility guarantees.

**Comment by gftea** [comment]
> > > there is no use case for using `:` as standalone formatter, and it is better the compiler to emit error to use : as standalone formatter because it is likely user want to use `:?` but miss `?` by mistake,
> 
> Unlikely to work, there's probably a crate out there that uses it, so it can't be removed due to backward compatibility guarantees.

can be a warnings?

**Comment by workingjubilee** [comment]
Yes, we can lint on that. In fact we already lint on a few bad `format_args!` inputs.
